NPE with Java Records:  Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.MethodBinding.original()" because "binding" is null

GWT Version: HEAD

--- 

When creating a record (code can be unused)

```java
public record TestRecord() {

  private static final int CAUSES_NPE = 1;

}
```
GWT compiler fails with exception
```
com.google.gwt.dev.jjs.InternalCompilerException: Error constructing Java AST
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.translateException(GwtAstBuilder.java:4239)
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.getInternalCompilerException(GwtAstBuilder.java:4678)
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.createMembers(GwtAstBuilder.java:4373)
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.processImpl(GwtAstBuilder.java:4171)
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.process(GwtAstBuilder.java:4213)
        at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater$UnitProcessorImpl.process(CompilationStateBuilder.java:128)
        at com.google.gwt.dev.javac.JdtCompiler$CompilerImpl.process(JdtCompiler.java:329)
        at org.eclipse.jdt.internal.compiler.Compiler.processCompiledUnits(Compiler.java:575)
        at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:475)
        at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:426)
        at com.google.gwt.dev.javac.JdtCompiler.doCompile(JdtCompiler.java:1028)
        at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater.compile(CompilationStateBuilder.java:322)
        at com.google.gwt.dev.javac.CompilationStateBuilder.doBuildFrom(CompilationStateBuilder.java:532)
        at com.google.gwt.dev.javac.CompilationStateBuilder.buildFrom(CompilationStateBuilder.java:464)
        at com.google.gwt.dev.cfg.ModuleDef.getCompilationState(ModuleDef.java:426)
        at com.google.gwt.dev.Precompile.precompile(Precompile.java:210)
        at com.google.gwt.dev.Precompile.precompile(Precompile.java:190)
        at com.google.gwt.dev.Precompile.precompile(Precompile.java:131)
        at com.google.gwt.dev.Compiler.compile(Compiler.java:192)
        at com.google.gwt.dev.codeserver.Recompiler.doCompile(Recompiler.java:362)
        at com.google.gwt.dev.codeserver.Recompiler.compile(Recompiler.java:175)
        at com.google.gwt.dev.codeserver.Recompiler.recompile(Recompiler.java:134)
        at com.google.gwt.dev.codeserver.Outbox.recompile(Outbox.java:135)
        at com.google.gwt.dev.codeserver.JobRunner.recompile(JobRunner.java:113)
        at com.google.gwt.dev.codeserver.JobRunner$2.run(JobRunner.java:90)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.MethodBinding.original()" because "binding" is null
        at com.google.gwt.dev.jjs.impl.ReferenceMapper.get(ReferenceMapper.java:99)
        at com.google.gwt.dev.jjs.impl.GwtAstBuilder.createMembers(GwtAstBuilder.java:4348)
        ... 27 more
```
Commenting the static field in the record makes the compiler happy again.

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -4337,18 +4337,21 @@ public class GwtAstBuilder {
 
       if (x.isRecord()) {
         // build implicit record component accessor methods, JDT doesn't declare them
-        for (JField field : type.getFields()) {
+        type.getFields().stream().filter(f -> !f.isStatic()).forEach(field -> {
           // Create a method binding that corresponds to the method we are creating, jdt won't
           // offer us one unless it was defined in source.
           char[] fieldName = field.getName().toCharArray();
           MethodBinding recordComponentAccessor = binding.getExactMethod(
                   fieldName, new TypeBinding[0], curCud.scope);
-
+          if (recordComponentAccessor == null) {
+            throw new InternalCompilerException(
+                "Missing method " + field.getName() + " for record type " + type.getName());
+          }
           // Get the record component, and pass on any annotations meant for the method
           JMethod componentMethod = typeMap.get(recordComponentAccessor);
           RecordComponentBinding component = binding.getRecordComponent(fieldName);
           processAnnotations(component.sourceRecordComponent().annotations, componentMethod);
-        }
+        });
 
         // At this time, we need to be sure a binding exists, either because the record declared
         // its own, or we make one specifically for it.

--- a/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -134,6 +134,13 @@ public class Java17Test extends GWTTestCase {
   }
 
   /**
+   * Record with a single static field.
+   */
+  record RecordWithStaticField() {
+    public static String foo = "bar";
+  }
+
+  /**
    * Record type that takes a record as a component
    */
   record RecordWithReferenceType(TopLevelRecord refType){}
@@ -164,6 +171,7 @@ public class Java17Test extends GWTTestCase {
     assertTrue(withValues.toString().contains("Banana"));
     assertEquals("Banana", withValues.name());
     assertEquals(7, withValues.count());
+    assertEquals("bar", RecordWithStaticField.foo);
     // Under the current implementation this next line would fail - this is not inconsistent with the spec,
     // but it is different than what the JVM does.
 //    assertEquals(0, new TopLevelRecord("", 0).hashCode());

